x-pack/plugins/apm/server/lib/services/get_service_dependencies/index.ts

```diff
@@ -0,0 +1,203 @@
+
+    const destination = destinationMap[spanDestination];
+
+    return {
+      destination: destination
+        ? destination
+        : {
+            [SPAN_DESTINATION_SERVICE_RESOURCE]: metricItem.key,
+          },
+      metrics: [metricItem],
+    };
+  }, []);
+
+  const metricsJoinedByDestination = joinByKey(
+    metricsWithMappedDestinations,
+    'destination',
+    (a, b) => ({
+      ...a,
+      ...b,
+      metrics: [...a.metrics, ...b.metrics],
+    })
+  );
+
+  const metricsByResolvedAddress = metricsJoinedByDestination.map(
+    ({ destination, metrics }) => {
+      const mergedMetrics = metrics.reduce<
+        Omit<ValuesType<typeof metrics>, 'key'>
+      >(
+        (prev, current) => {
+          return {
+            value: {
+              count: prev.value.count + current.value.count,
+              latency_sum: prev.value.latency_sum + current.value.latency_sum,
+              error_count: prev.value.error_count + current.value.error_count,
+            },
+            timeseries: joinByKey(
+              [...prev.timeseries, ...current.timeseries],
+              'x',
+              (a, b) => ({
+                x: a.x,
+                count: a.count + b.count,
+                latency_sum: a.latency_sum + b.latency_sum,
+                error_count: a.error_count + b.error_count,
+              })
+            ),
+          };
+        },
+        {
+          value: {
+            count: 0,
+            latency_sum: 0,
+            error_count: 0,
+          },
+          timeseries: [],
+        }
+      );
+
+      const deltaAsMinutes = (end - start) / 60 / 1000;
+
+      const destMetrics = {
+        latency: {
+          value:
+            mergedMetrics.value.count > 0
+              ? mergedMetrics.value.latency_sum / mergedMetrics.value.count
+              : null,
+          timeseries: mergedMetrics.timeseries.map((point) => ({
+            x: point.x,
+            y: point.count > 0 ? point.latency_sum / point.count : null,
+          })),
+        },
+        throughput: {
+          value:
+            mergedMetrics.value.count > 0
+              ? mergedMetrics.value.count / deltaAsMinutes
+              : null,
+          timeseries: mergedMetrics.timeseries.map((point) => ({
+            x: point.x,
+            y: point.count > 0 ? point.count / deltaAsMinutes : null,
+          })),
+        },
+        error_rate: {
+          value:
+            mergedMetrics.value.count > 0
+              ? (mergedMetrics.value.error_count ?? 0) /
+                mergedMetrics.value.count
+              : null,
+          timeseries: mergedMetrics.timeseries.map((point) => ({
+            x: point.x,
+            y: point.count > 0 ? (point.error_count ?? 0) / point.count : null,
+          })),
+        },
+      };
+
+      if ('service' in destination) {
+        return {
+          name: destination.service!.name,
+          serviceName: destination.service!.name,
```

Comment:
Would be better if we could avoid forcing the compiler (using `!.`)
I'm actually surprised this is necessary - doesn't typescript narrow the type inside the `if` block?

Reply:
I was surprised as well, it might be because of the way `joinByKey` returns its type. Will have a look, and I can possibly add `&& destination.service` here (though I think I would prefer an error here).